### ISSUE TITLE
Gulp task never exits

### ISSUE THREAD

**Issue by Coaxial** [SBoudrias/gulp-istanbul#92]
When I run the coverage task in gulp with mocha and istanbul, gulp reports the task as done. However, it doesn't exit.

I would have expected gulp to exit once the task is over.

gulpfile.js:
```javascript
'use strict';

const gulp = require('gulp');
const mocha = require('gulp-spawn-mocha');
// istanbul requires the non-threaded gulp-mocha version
const mocha_istanbul = require('gulp-mocha');
const istanbul = require('gulp-istanbul');
const eslint = require('gulp-eslint');
const runSequence = require('run-sequence');

const test_files = 'test/**/*_test.js';
const test_helper_file = 'test/test_helper.js';
const lib_files = 'lib/**/*.js';
const app_file = 'index.js';
const min_coverage = 90;

// ...
// other unrelated tasks
// ...

gulp.task('pre-cov', () => {
  return gulp.src([lib_files])
    .pipe(istanbul({ includeUntested: true }))
    .pipe(istanbul.hookRequire());
});

gulp.task('cov', ['pre-cov'], () => {
  process.env.NOCK_BACK_MODE = 'lockdown';
  return gulp.src([test_files], { read: false })
    .pipe(mocha_istanbul({ reporter: 'dot' }))
    .pipe(istanbul.writeReports())
    .pipe(istanbul.enforceThresholds({ thresholds: { global: min_coverage } }));
});
```

Output when running the task (the `^C` is me exiting gulp since it doesn't do it itself)
```
$ gulp cov
[13:26:20] Using gulpfile ~/project/gulpfile.js
[13:26:20] Starting 'pre-cov'...
[13:26:21] Finished 'pre-cov' after 125 ms
[13:26:21] Starting 'cov'...


  ․․․․․․․․․․․․․․․․․․․․․․․

  23 passing (249ms)

---------------------|----------|----------|----------|----------|----------------|
File                 |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
---------------------|----------|----------|----------|----------|----------------|
 lib/                |      100 |      100 |      100 |      100 |                |
  my_lib.js          |      100 |      100 |      100 |      100 |                |
---------------------|----------|----------|----------|----------|----------------|
All files            |      100 |      100 |      100 |      100 |                |
---------------------|----------|----------|----------|----------|----------------|


=============================== Coverage summary ===============================
Statements   : 100% ( 136/136 )
Branches     : 100% ( 25/25 )
Functions    : 100% ( 20/20 )
Lines        : 100% ( 135/135 )
================================================================================
[13:26:21] Finished 'cov' after 921 ms
^C
```

All other gulp tasks running fine and exiting properly.
```bash
$ gulp -v
[13:30:28] CLI version 3.9.1
[13:30:28] Local version 3.9.1
$ node -v
v4.4.3
```

**Comment by SBoudrias** [comment]
That's not related to gulp-istanbul. Some of your other task is hanging.

If you see the reports being printed, that mean gulp-istanbul correctly finished.

**Comment by miparnisari** [comment]
I'm experiencing the same with the following task `tests`:

```javascript
var gulp = require('gulp');
var istanbul = require('gulp-istanbul');
var mocha = require('gulp-mocha');

gulp.task('pre-test', function () {
  return gulp.src(['server.js'])
    .pipe(istanbul())
    .pipe(istanbul.hookRequire());
});

gulp.task('tests', ['pre-test'], function () {
  return gulp.src(['test/*.js'])
    .pipe(mocha({ reporter: 'mocha-pretty-bunyan-nyan' }))
    .pipe(istanbul.writeReports({ dir: './coverage', reporters: ['html', 'text-summary'] }))
});
```

My versions are as follows:
```
"mocha": "^3.2.0",
"gulp-istanbul": "^1.1.1",
"gulp-mocha": "^3.0.1"
```

**Comment by Coaxial** [comment]
@miparnisari Sorry, I never found the root cause for that issue

**Comment by miparnisari** [comment]
I've got another project with the exact same tasks and it does exit. It must be something misconfigured within the projec.t.

**Comment by Ks89** [comment]
I have the same problem also with gulp 4 alpha.

```
=============================== Coverage summary ===============================
Statements   : 90.21% ( 986/1093 )
Branches     : 89.26% ( 399/447 )
Functions    : 91.14% ( 72/79 )
Lines        : 90.21% ( 986/1093 )
================================================================================
[13:39:57] Finished 'testInternal' after 1.22 min
[13:39:57] Finished 'test' after 1.22 min
```

Everything is ok, my tests are ok, but my gulp process is blocked here. 

Locally is not a big problem, but on Travis CI and other services yes.

**Comment by SBoudrias** [comment]
@Ks89 so if you remove gulp-istanbul, your process ends properly?

Also, is everyone reporting this issue using gulp 4 alpha?

**Comment by Ks89** [comment]
Ok I found the problem. gulp-istanbul is ok

I'm doing integration testing of my APIs and it seems that **mongoose connections are creating problems to gulp-mocha.**
With unit testing everything is ok, because I have a direct reference to mongoose object and after every test I call 'mongoose.disconnect()'.
With integration testing, I'm using supertest, so I'm not creating the mongoose connection in my test but in the main app.js. I'll continue my investigations to find a real fix.

Searching on the web I discovered that gulp-mocha + mongoose create problems to many people.
The issue is always related to mongoose connections.

The real problem is how to disconnect mongoose at the end of every test with supertest. Mongoose.disconnect() is not working in this situation.

**Comment by Coaxial** [comment]
@Ks89 I didn't use any Mongoose dependency when I experienced the issue.

@SBoudrias I had the issue with an earlier version of Gulp, haven't tried with version 4.

**Comment by Ks89** [comment]
Weird. 
I think that my problem is caused by mongoose connections, but probably this issue happens every time you have something "not terminated", for instance connections or a piece of code that is waiting an event.
I don't know, it's only a supposition.

@Coaxial Did you try to execute only one test (one file) specifying the entire file path? I found my problem (files related to integration testing with mongoose) with this technique.

**Comment by Ks89** [comment]
This issue is similar: https://github.com/sindresorhus/gulp-mocha/issues/118

**Comment by SBoudrias** [comment]
The thing that comes out here in all cases, the process wasn't terminating because of bugs and issues in your setup unrelated to gulp-istanbul.

These issues are all different because you don't all have the same project.

**Comment by Ks89** [comment]
Yep, I'm agree with you. gulp-istanbul is ok. 
Thank u for your support.

I increased mocha's timeout to 10000 ms and I used gulp-exit in this way:

```
.pipe(mocha({
      timeout: 10000
    }))
.pipe(exit());
```
And my build on TravisCI is ok.

The funny thing is that a plugin like "gulp-exit" exists. Ok, our code could be bad, but probably also gulp-mocha's developers should manage this in a better way. It's a really common problem without a solution.

Thank u again,
Ks89
